base/collection_n1ql.go

```diff
@@ -0,0 +1,125 @@
+package base
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/couchbase/gocb"
+	sgbucket "github.com/couchbase/sg-bucket"
+	pkgerrors "github.com/pkg/errors"
+)
+
+var _ N1QLStore = &Collection{}
+
+// Keyspace for a collection is bucket name until wider collection support is added
+func (c *Collection) Keyspace() string {
+	return c.Bucket().Name()
+}
+
+func (c *Collection) Query(statement string, params map[string]interface{}, consistency ConsistencyMode, adhoc bool) (resultsIterator sgbucket.QueryResultIterator, err error) {
+
+	bucketStatement := strings.Replace(statement, KeyspaceQueryToken, c.Bucket().Name(), -1)
+
+	n1qlOptions := &gocb.QueryOptions{
+		ScanConsistency: gocb.QueryScanConsistency(consistency),
+		Adhoc:           adhoc,
+		NamedParameters: params,
+	}
+
+	waitTime := 10 * time.Millisecond
+	for i := 1; i <= MaxQueryRetries; i++ {
+		Tracef(KeyQuery, "Executing N1QL query: %v", UD(bucketStatement))
+		queryResults, queryErr := c.cluster.Query(bucketStatement, n1qlOptions)
+		if queryErr == nil {
+			resultsIterator := &gocbRawIterator{
+				rawResult: queryResults.Raw(),
+			}
+			return resultsIterator, queryErr
+		}
+
+		// Timeout error - return named error
+		if errors.Is(queryErr, gocb.ErrTimeout) {
+			return resultsIterator, ErrViewTimeoutError
+		}
+
+		// Non-retry error - return
+		if !isTransientIndexerError(queryErr) {
+			Warnf("Error when querying index using statement: [%s] parameters: [%+v] error:%v", UD(bucketStatement), UD(params), queryErr)
+			return resultsIterator, pkgerrors.WithStack(queryErr)
+		}
+
+		// Indexer error - wait then retry
+		err = queryErr
+		Warnf("Indexer error during query - retry %d/%d after %v.  Error: %v", i, MaxQueryRetries, waitTime, queryErr)
+		time.Sleep(waitTime)
+
+		waitTime = waitTime * 2
+	}
+
+	Warnf("Exceeded max retries for query when querying index using statement: [%s], err:%v", UD(bucketStatement), err)
```

Comment:
Params + consistency with non-retry warning above

```suggestion
	Warnf("Exceeded max retries for query when querying index using statement: [%s] parameters: [%+v] error:%v", UD(bucketStatement), UD(params), err)
```